records/providers_manager.go

```diff
@@ -0,0 +1,413 @@
+	if err != nil {
+		return nil, err
+	}
+	pm.cache = cache
+	pm.cleanupInterval = defaultCleanupInterval
+	if err := pm.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+	pm.ctx, pm.cancel = context.WithCancel(ctx)
+	pm.run()
+	return pm, nil
+}
+
+func (pm *ProviderManager) run() {
+	pm.wg.Add(1)
+	go func() {
+		defer pm.wg.Done()
+
+		var gcQuery dsq.Results
+		gcTimer := time.NewTimer(pm.cleanupInterval)
+
+		defer func() {
+			gcTimer.Stop()
+			if gcQuery != nil {
+				gcQuery.Close()
+			}
+			if err := pm.dstore.Flush(context.Background()); err != nil {
+				log.Error("failed to flush datastore: ", err)
+			}
+		}()
+
+		var gcQueryRes <-chan dsq.Result
+		var gcSkip map[string]struct{}
+		var gcTime time.Time
+		for {
+			select {
+			case np := <-pm.newprovs:
+				err := pm.addProv(np.ctx, np.key, np.val)
+				if err != nil {
+					log.Error("error adding new providers: ", err)
+					continue
+				}
+				if gcSkip != nil {
+					// we have an gc, tell it to skip this provider
+					// as we've updated it since the GC started.
+					gcSkip[mkProvKeyFor(np.key, np.val)] = struct{}{}
+				}
+			case gp := <-pm.getprovs:
+				provs, err := pm.getProvidersForKey(gp.ctx, gp.key)
+				if err != nil && err != ds.ErrNotFound {
+					log.Error("error reading providers: ", err)
+				}
+
+				// set the cap so the user can't append to this.
+				gp.resp <- provs[0:len(provs):len(provs)]
+			case res, ok := <-gcQueryRes:
+				if !ok {
+					gcQuery.Close()
+					gcTimer.Reset(pm.cleanupInterval)
+
+					// cleanup GC round
+					gcQueryRes = nil
+					gcSkip = nil
+					gcQuery = nil
+					continue
+				}
+				if res.Error != nil {
+					log.Error("got error from GC query: ", res.Error)
+					continue
+				}
+				if _, ok := gcSkip[res.Key]; ok {
+					// We've updated this record since starting the
+					// GC round, skip it.
+					continue
+				}
+
+				// check expiration time
+				t, err := readTimeValue(res.Value)
+				switch {
+				case err != nil:
+					// couldn't parse the time
+					log.Error("parsing providers record from disk: ", err)
+					fallthrough
+				case gcTime.Sub(t) > ProvideValidity:
+					// or expired
+					err = pm.dstore.Delete(pm.ctx, ds.RawKey(res.Key))
+					if err != nil && err != ds.ErrNotFound {
+						log.Error("failed to remove provider record from disk: ", err)
+					}
+				}
+
+			case gcTime = <-gcTimer.C:
+				// You know the wonderful thing about caches? You can
+				// drop them.
+				//
+				// Much faster than GCing.
+				pm.cache.Purge()
+
+				// Now, kick off a GC of the datastore.
+				q, err := pm.dstore.Query(pm.ctx, dsq.Query{
+					Prefix: ProvidersKeyPrefix,
+				})
+				if err != nil {
+					log.Error("provider record GC query failed: ", err)
+					continue
+				}
+				gcQuery = q
+				gcQueryRes = q.Next()
+				gcSkip = make(map[string]struct{})
+			case <-pm.ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (pm *ProviderManager) Close() error {
+	pm.cancel()
+	pm.wg.Wait()
```

Comment:
Use channel instead.
```suggestion
	<-pm.closed
```